### DISCUSSION TITLE
Allow response headers to be injected within dependencies

### DISCUSSION THREAD

**Discussion by soundstripe** [fastapi/fastapi#9052] [Questions] [answered]
### Is your feature request related to a problem

I'd like to be able inject a response header from inside a dependency, as we can from path operation functions. The dependency injection system should operate the same for dependency functions as for path operation functions.

Example code:

```
from datetime import datetime

from fastapi import FastAPI, Depends
from starlette.responses import Response
from starlette.testclient import TestClient

app = FastAPI()


def cache_control(response: Response):
    response.headers.setdefault('last-modified', datetime.now().isoformat())


def other_dependency():
    return {'foo': 'bar'}


@app.get('/doc', dependencies=[Depends(cache_control)])
def get_doc(other: dict = Depends(other_dependency)):
    document = {'other': other}
    return document


if __name__ == '__main__':
    client = TestClient(app=app)
    print(client.get('/doc').headers)
```

The result of the above code is that the last-modified header is duplicated because the dependency resolver expects that only the path operation function will try to add headers, and only once. Adding additional dependencies causes exponential duplication.

Headers get merged here:
https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L489

At that point in the above scenario, `response is sub_response` is `True` -- they are the same object.

**Comment by denisSurkov** [comment]
Hello 
As I understand [middleware](https://fastapi.tiangolo.com/tutorial/middleware/) perfectly fit your needs

**Comment by soundstripe** [comment]
My use case is for generating a last-modified or etag header which relies on looking up database objects, potentially filtered and/or paginated just like the path operation. For dependency injection and code reuse reasons it seems like it would be easier to perform this inside the app.

**Comment by alarsymstu** [comment]
```python
from fastapi import Depends, APIRouter, Response, FastAPI


router = APIRouter()


async def auth_header(response: Response):
    response.set_cookie('Authorization', 'loh')
    return response


async def empty_depends():
    return


@router.get('/')
async def main(updated_resp: Response = Depends(auth_header),
               empty: None = Depends(empty_depends)):
    return 'updated_resp'


app = FastAPI()
app.include_router(router)
```

Same issue. **Authorization** header is duplicated. Currently we solve this problem with middleware, but it's strange behavior and i expect to see just one in any situation even if it's list rather then dictionary.

Perhaps, in other frameworks we can see that headers not duplicated and represented as dict

**Comment by soundstripe** [comment]
FastAPI should allow multiple headers with the same name (as per the RFC - https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2) but should not duplicate them when they are specified only once.

**Comment by sergey-babayan-v** [comment]
It could help to write more clean code, with injections, middlewares are too heavy for such tasks.

**Comment by erichaydel** [comment]
Adding a +1 to this. For my enterprise workflow, we need to send logging to an external server. Part of those logs are the response codes we send back to the user. Using a middleware would be heavy, but also hard to implement cleanly because we need different sorts of logs depending on the endpoint.  (Some endpoints have users we have to track, some have database accesses, etc. which need to go to different places). 

This would be a really nice feature for us to make our code more readable and cleaner to test.

**Comment by YuriiMotov** [comment, marked as answer]
Seems that was fixed somewhere.
In current version (0.115.14) the code example prints:
```
Headers({'content-length': '23', 'content-type': 'application/json', 'last-modified': '2025-07-03T15:58:38.314410'})
```
